### ISSUE TITLE
Relativistic Radiative Transfer in arbitrary spacetime

### ISSUE THREAD

**Issue by sbhkmr1999** [einsteinpy/einsteinpy#339]
🎯 **Goal**

Solving the Relativistic Radiative Transfer first in Schwarzschild and Kerr spacetime and then for arbitrary spacetime, for Emission Intensity (Emissivity) , for an observer in equatorial plane  (for starters) and then at any arbitrary angle.

💡 **Possible solutions**

https://arxiv.org/abs/1801.10452
https://arxiv.org/abs/1602.03184
http://discovery.ucl.ac.uk/1426969/1/Younsi_Ziri_741337%20_Ziri_Thesis.pdf

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

**Comment by sbhkmr1999** [comment]
Some theoretical literature on the relativistic radiative transfer equations:

http://articles.adsabs.harvard.edu//full/2008ChJAS...8..226W/0000226.000.html
https://arxiv.org/abs/1304.5691
https://arxiv.org/abs/1207.4234

**Comment by ritzvik** [comment]
@Relativist1 Thanks for the issue. Looking forward to adding this to EPY.

**Comment by Bhavam** [comment]
Hi all , so apparently going through the requirements of the radiative transfer equation turns out it requires a bunch of parameters in the spacetime objects to be implemented for eg: lorentz invariant emission and  absorption coefficients which i did not find in the related classes , can @Relativist1 or @ritzvik verify this problem so I can start working on something to help on this issue.

**Comment by ritzvik** [comment]
Yups, you’re correct. No such classes are available as of now. Is you’re willing to go ahead with this “sort-of” mini-project, you would need to create the helper functions and classes along with it. Ofc, I’ll be there for assistance. Also, @Relativist1 , can you provide some related papers??

Regards,
Ritwik

**Comment by Bhavam** [comment]
@ritzvik okay cool ill start with defining the required class methods and parameters for schwarzschild bodies initially and keep you posted with a PR.

**Comment by Relativist1** [comment]
https://arxiv.org/abs/1207.4234
Here is what I found that can be done.

**Comment by Bhavam** [comment]
Hey @ritzvik very sorry for the delay ,  so I've outlined some basic parameters based on RAPTOR1 that we have to define for radiative transfer those being -
1.`4 velocity` comprising of the wave functions of the photons in list along with a helper method to compute it from photons position/velocity vector
2. `differential step size` comprising of an appropriate step size based on position or velocity vectors or differential lambda 
3. `frequencies` of the photons 
3.`intensity field` vector and a separate variable for current intensity value 
4.`tau` vector

So basically after figuring out which of these params exist already in the codebase should I start defining those which don't  in a new class ? or in each spacetime class. 

Also apparently for runge kutta , epy uses RK45 function but im a bit confused on how to use it could you clarify `fun` parameter used in it ?

**Comment by ritzvik** [comment]
> Hey @ritzvik very sorry for the delay , so I've outlined some basic parameters based on RAPTOR1 that we have to define for radiative transfer those being -
> 1.`4 velocity` comprising of the wave functions of the photons in list along with a helper method to compute it from photons position/velocity vector
> 2. `differential step size` comprising of an appropriate step size based on position or velocity vectors or differential lambda
> 3. `frequencies` of the photons
> 3.`intensity field` vector and a separate variable for current intensity value
> 4.`tau` vector
> 
> So basically after figuring out which of these params exist already in the codebase should I start defining those which don't in a new class ? or in each spacetime class.
> 
> Also apparently for runge kutta , epy uses RK45 function but im a bit confused on how to use it could you clarify `fun` parameter used in it ?

I didn't go through the paper, I will go through it at a later stage. Regarding `fun` parameter, it is used to pass a function(which may be a vector function) to calculate `f(x)` given `x`(which may be vector, again). This is because of differential equation `y' = f(x)`, each new `y` is calculated by `y=y+h*f(x)`, where `h` is the step-size.
This is the simplest way of calculating solutions of ODE numerically, although RK45 involves a much more complicated algorithm.

**Comment by Relativist1** [comment]
* https://arxiv.org/pdf/2003.06943.pdf
*  https://arxiv.org/pdf/1909.10322.pdf
#105

**Comment by GTron-1729** [comment]
So as per our discussion @Relativist1 the equations after some manipulations can be interpreted as integration of [](https://drive.google.com/file/d/133lS2RoNeF2TKNqKw_enlvJDvXDepHGI/view?usp=sharing) where b^2 is expressed in terms of two parameters as { n + e^2 } and the main problem bottles down to finding the upper limit of this integral where the lower bound indicates the point at observer and the upper bound indicates the the point from where the photon was emitted from he accretion disks. I have solved the problem in 3 lines of code using matlab 

    fun = @(r)1./((r.^2).*((1/100)-(1/r.^2).*(1-(2./r)))^1/2);
    G1 = @(r)(fun(r) - 0.05);
    pLB = fzero(G1,1);pLB
                                              Where i have assumed the value of M=1  and b=10.
The output can be found here [](https://drive.google.com/file/d/1Xsi1Ytpmn-LgX0B_isU8lI2tvT8t6xZL/view?usp=sharing) . where the circled value is the value of upper bound. 

Besides I am also working on python implementation of it where we would write certain functions to find the upper bound. The first function would be to calculate the area between the lower and upper bounds. The second function would be ( the first function - the numerical value ) of integral that is known and the roots to this equation would be the values of upper bound.

**Comment by GTron-1729** [comment]
So as per our discussion @Relativist1 the equations after some manipulations can be interpreted as integration of https://drive.google.com/file/d/133lS2RoNeF2TKNqKw_enlvJDvXDepHGI/view?usp=sharing where b^2 is expressed in terms of two parameters as { n + e^2 } and the main problem bottles down to finding the upper limit of this integral where the lower bound indicates the point at observer and the upper bound indicates the the point from where the photon was emitted from he accretion disks. I have solved the problem in 3 lines of code using matlab

fun = @(r)1./((r.^2).*((1/100)-(1/r.^2).*(1-(2./r)))^1/2);
G1 = @(r)(fun(r) - 0.05);
pLB = fzero(G1,1);pLB
                                          Where i have assumed the value of M=1  and b=10.
The output can be found here  https://drive.google.com/file/d/1Xsi1Ytpmn-LgX0B_isU8lI2tvT8t6xZL/view?usp=sharing . where the circled value is the value of upper bound.

Besides I am also working on python implementation of it where we would write certain functions to find the upper bound. The first function would be to calculate the area between the lower and upper bounds. The second function would be ( the first function - the numerical value ) of integral that is known and the roots to this equation would be the values of upper bound.

**Comment by GTron-1729** [comment]
Now i have done it in python too @Relativist1  https://drive.google.com/file/d/1qEXRIDpgitWe-6LT2CEuAVYA-Xhj-Neh/view?usp=sharing  --> this is the code along with the output.
